### PR TITLE
Bluetooth Serial JSON Control and WiFi Deferment

### PR DESCRIPTION
**Pull request by ChuckMash**
Due to a very similar PR (#2925) also today, I will open this a draft to encourage discussion between the 2 similar implementations.

* Allows access to JSON API over Bluetooth Serial Connection.
* __Normal WiFi functionality is disabled while Bluetooth is in use. OTA disabled due to size constraints__
* Bluetooth device shows up as whatever is set in "Server description", "WLED" by default.
* Normal WiFi functionality can be accessed for general WLED setup under the following conditions
  * Normal WiFi functionality until "Apply preset at boot" is set.
  * If active preset changes by any means during first 5 seconds, normal WiFi functionality will resume. (simple button press would suffice)
  * If 'e' is sent over the bluetooth serial connection, normal WiFi functionality will resume.
* [Serial Bluetooth Terminal](https://play.google.com/store/apps/details?id=de.kai_morich.serial_bluetooth_terminal) Android app for testing
* [Reference: Python to ESP32 bluetooth serial](https://medium.com/@18218004/devlog-6-bluetooth-and-esp32-ba076a8e207d)
* [Example Python Script](https://gist.github.com/ChuckMash/7752b77ea1bf204a7e8f0acac6e27641) for WLED control over Bluetooth Serial

### CODE REVIEW

File: tools/WLED_ESP32_NO_OTA.csv [hunk 1]
```diff
@@ -0,0 +1,5 @@
+# Name,   Type, SubType, Offset,  Size, Flags
```
Comment (blazoncek) [inline tools/WLED_ESP32_NO_OTA.csv:1]: Is the code size >1.5MB?

File: usermods/bluetooth_serial/bluetooth_serial.h [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+    void setup() {
+
+      // Tell WLED to defer all WiFi Connectivity
+      deferConnections = true;
+      
+      // Begin Bluetooth device with same name as server description
+      SerialBT.begin(serverDescription);
+
+      Serial.println("Bluetooth Serial Usermod setup complete");
```
Comment (blazoncek) [inline usermods/bluetooth_serial/bluetooth_serial.h:21]: Use DBUG_PRINTLN() and use flash strings F()

File: usermods/bluetooth_serial/bluetooth_serial.h [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+      Serial.println();
+    }
+
+
+    // Cleans up and lets WLED run WiFi normally.
+    // Once called, this usermod is disabled until next boot
+    void stop(){
+      SerialBT.flush();
+      SerialBT.end();
+      deferConnections=false;
```
Comment (blazoncek) [inline usermods/bluetooth_serial/bluetooth_serial.h:45]: Perhaps also add `forceReconnect = true;`.

File: usermods/bluetooth_serial/bluetooth_serial.h [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    void loop() {
+
+      // If Wifi Deferment has ended, do nothing
+      if(!deferConnections){
+        return;
+      }
+
+      // If not set up or if preset changes in first 5* seconds of boot, disable Bluetooth and end WiFi deferment
+      unsigned long now = millis();
+      if(now > 1000 && now < 6000 && (currentPreset != bootPreset || currentPreset == 0 && bootPreset == 0)){
```
Comment (blazoncek) [inline usermods/bluetooth_serial/bluetooth_serial.h:59]: Please add parentheses to clear ambiguous condition.

**Review by blazoncek — CHANGES_REQUESTED** [review]
I am wondering how do you pair? If it is needed.
Otherwise minor issues but I would like them fixed.

File: tools/WLED_ESP32_NO_OTA.csv (hunk 1 above)
Comment (ChuckMash) [reply tools/WLED_ESP32_NO_OTA.csv:1]: > Is the code size >1.5MB?

Yes, 1.8MB+ with the compile options in the usermod readme

**Comment by ChuckMash** [comment]
> I am wondering how do you pair? If it is needed.

For example if you would want to use this to access the JSON API from an Android device you would find the device under available devices with the WLED server description name, the pairing is automatic from there.

File: tools/WLED_ESP32_NO_OTA.csv (hunk 1 above)
Comment (blazoncek) [reply tools/WLED_ESP32_NO_OTA.csv:1]: No OTA will be a hard sell.

**Comment by blazoncek** [comment]
> For example if you would want to use this to access the JSON API from an Android device you would find the device under available devices with the WLED server description name, the pairing is automatic from there.

So something like mouse or keyboard? I would still like to see some pairing option with PIN if possible.

**Comment by ChuckMash** [comment]
>  I would still like to see some pairing option with PIN if possible.

Agreed, I've been looking into it

File: usermods/bluetooth_serial/bluetooth_serial.h (hunk 2 above)
Comment (ChuckMash) [reply usermods/bluetooth_serial/bluetooth_serial.h:21]: Is your thinking here to replace all serial out with DEBUG equivalent or just this message?
Getting things like the bluetooth device name and hardware address can be useful, but I can see how they may belong as a debug message instead.

Comment (blazoncek) [reply usermods/bluetooth_serial/bluetooth_serial.h:21]: Avoid using Serial.print() for printing debug messages. Use it only if you intend output to Serial interface.
Debug can go to networked recipient.

**Comment by ChuckMash** [comment]
That should deal with the bulk of the talking points so far, with exception of the pairing process which I am still looking into.

File: wled00/wled.cpp [hunk 5]
```diff
@@ -716,6 +716,10 @@ void WLED::initInterfaces()
 
 void WLED::handleConnection()
 {
```
[the remark this answers is not in the document]
Comment (johne) [reply wled00/wled.cpp:718]: tried it on a wt32-eth01 device... it didn't work... I set up 2 characteristics per json endpoint and it gets through about 1/2 of them before rebooting... i'm guessin the eth stack takes up some memory which doesn't leave enough for my other characteristics... I could route all calls through 1 set of characteristics and just include something in the json to id what endpoint it is in response to... that might help save a bit on memory (if that is the problem)... I might put in some debug prints to show the memory growing as I allocate characteristics to confirm that memory is the problem

**Comment by ChuckMash** [comment]
Stalebot reminded me, I don't have the time at the moment to finish this up.
Back when I was working on it there was upstream issues with bluetooth serial and being able to set a pairing pin.
Some time later I may re-open this draft and continue.

**Pull request closed without merging**